File: 07_midas_aruco_fusion/core/anypoint_controller.py
"""
core/anypoint_controller.py — Mouse-driven anypoint view controller
====================================================================
Melampirkan mouse callback ke jendela OpenCV agar user bisa menggeser
sudut pandang Moildev secara interaktif:

  • Drag kiri-kanan  → yaw   (geser horizontal)
  • Drag atas-bawah  → pitch (geser vertikal)
  • Scroll wheel     → zoom  (dekat / jauh)
  • Tekan R          → reset ke nilai awal
  • Tekan S          → print nilai saat ini ke terminal (untuk disimpan)

Kontrol ini aktif di semua mode (kalibrasi maupun live).
Dipanggil dari run_fusion.py, hanya jika --fisheye aktif.
"""

import cv2


class AnypointController:
    """
    Mengelola interaksi mouse untuk mengontrol parameter Moildev anypoint.

    Parameters
    ----------
    undistorter : MoilUndistorter
        Instance MoilUndistorter yang akan dikontrol.
    sensitivity_deg_per_px : float
        Seberapa banyak derajat berubah per piksel drag (default 0.15°/px).
    zoom_step : float
        Perubahan zoom per scroll tick (default 0.1).
    """

    def __init__(
        self,
        undistorter,
        sensitivity_deg_per_px: float = 0.15,
        zoom_step: float = 0.1,
    ):
        self._u             = undistorter
        self._sens          = sensitivity_deg_per_px
        self._zoom_step     = zoom_step

        # simpan nilai awal untuk reset
        self._init_pitch    = undistorter.pitch
        self._init_yaw      = undistorter.yaw
        self._init_roll     = undistorter.roll
        self._init_zoom     = undistorter.zoom

        # state mouse drag
        self._dragging      = False
        self._drag_start_x  = 0
        self._drag_start_y  = 0
        self._pitch_at_drag = 0.0
        self._yaw_at_drag   = 0.0

        # flag agar update_maps tidak dipanggil terlalu sering saat drag
        self._dirty         = False

    # ── Callback utama ──────────────────────────────────────────────────────

    def mouse_callback(self, event, x, y, flags, param):
        """Dipasang ke jendela cv2 via cv2.setMouseCallback."""

        if event == cv2.EVENT_LBUTTONDOWN:
            self._dragging      = True
            self._drag_start_x  = x
            self._drag_start_y  = y
            self._pitch_at_drag = self._u.pitch
            self._yaw_at_drag   = self._u.yaw

        elif event == cv2.EVENT_MOUSEMOVE and self._dragging:
            dx = x - self._drag_start_x     # geser horizontal → yaw
            dy = y - self._drag_start_y     # geser vertikal   → pitch
            new_pitch = self._pitch_at_drag - dy * self._sens
            new_yaw   = self._yaw_at_drag   + dx * self._sens
            self._u.update_maps(pitch=new_pitch, yaw=new_yaw)
            self._dirty = True

        elif event == cv2.EVENT_LBUTTONUP:
            self._dragging = False

        elif event == cv2.EVENT_MOUSEWHEEL:
            # flags > 0  → scroll up  → zoom in (nilai naik)
            # flags < 0  → scroll down→ zoom out
            delta = self._zoom_step if flags > 0 else -self._zoom_step
            self._u.update_maps(zoom=self._u.zoom + delta)
            self._dirty = True

    def reset(self):
        """Reset ke nilai awal saat MoilUndistorter diinisiasi."""
        self._u.update_maps(
            pitch=self._init_pitch,
            yaw=self._init_yaw,
            roll=self._init_roll,
            zoom=self._init_zoom,
        )
        self._dirty = True

    # ── Overlay HUD ─────────────────────────────────────────────────────────

    def draw_overlay(self, frame):
        """
        Gambar panel info Anypoint di pojok kanan bawah frame.
        Kembalikan frame yang sudah diberi anotasi (tidak mengubah original).
        """
        h, w = frame.shape[:2]
        x0, y0 = w - 260, h - 100
        x1, y1 = w - 8,   h - 8

        cv2.rectangle(frame, (x0, y0), (x1, y1), (10, 10, 30), -1)
        cv2.rectangle(frame, (x0, y0), (x1, y1), (0, 180, 255), 1)

        def put(text, row, color=(200, 230, 255)):
            cv2.putText(frame, text, (x0 + 6, y0 + 18 + row * 20),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.38, color, 1, cv2.LINE_AA)

        put("[ ANYPOINT CTRL ]", 0, (0, 200, 255))
        put(f"Pitch : {self._u.pitch:+.1f}deg  (drag up/dn)", 1)
        put(f"Yaw   : {self._u.yaw:+.1f}deg  (drag L/R)", 2)
        put(f"Zoom  : {self._u.zoom:.2f}x  (scroll)", 3)
        put("R=reset  S=print params", 4, (100, 200, 100))

        return frame

    def attach(self, window_name: str):
        """Pasang callback ke jendela OpenCV yang sudah dibuat."""
        cv2.setMouseCallback(window_name, self.mouse_callback)

File: module/volume/src/fisheye/anypoint_controller.cpp
/**
 * @file anypoint_controller.cpp
 * @brief Anypoint controller implementation.
 */

#include "anypoint_controller.h"
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <cmath>
#include <sstream>
#include <iomanip>

namespace fusion {

AnypointController::AnypointController(double init_pitch, double init_yaw,
                                       double init_zoom)
    : pitch_(init_pitch), yaw_(init_yaw), zoom_(init_zoom),
      init_pitch_(init_pitch), init_yaw_(init_yaw), init_zoom_(init_zoom)
{}

void AnypointController::mouse_callback(int event, int x, int y, int flags) {
    (void)flags;
    if (event == cv::EVENT_LBUTTONDOWN) {
        dragging_ = true;
        drag_start_x_ = x;
        drag_start_y_ = y;
        drag_start_pitch_ = pitch_;
        drag_start_yaw_ = yaw_;
    } else if (event == cv::EVENT_MOUSEMOVE && dragging_) {
        double dx = x - drag_start_x_;
        double dy = y - drag_start_y_;
        pitch_ = drag_start_pitch_ + dy * 0.1;
        yaw_ = drag_start_yaw_ + dx * 0.1;
        pitch_ = std::clamp(pitch_, -75.0, 75.0);
        yaw_ = std::clamp(yaw_, -75.0, 75.0);
        changed_ = true;
    } else if (event == cv::EVENT_LBUTTONUP) {
        dragging_ = false;
    } else if (event == 10 /* cv::EVENT_MOUSEWHEEL */) {
        // Mocking cv::EVENT_MOUSEWHEEL since highgui is not linked
        // In actual GTK use case, we use scroll_event
        double delta = (flags > 0) ? 0.1 : -0.1;
        zoom_ = std::clamp(zoom_ + delta, 1.0, 8.0);
        changed_ = true;
    }
}

void AnypointController::attach(const std::string& window_name) {
    (void)window_name;
    // Removed cv::setMouseCallback to drop highgui dependency
}

void AnypointController::reset() {
    pitch_ = init_pitch_;
    yaw_ = init_yaw_;
    zoom_ = init_zoom_;
    changed_ = true;
}

void AnypointController::draw_overlay(cv::Mat& frame) const {
    int bw = 220, bh = 80;
    int bx = frame.cols - bw - 10, by = frame.rows - bh - 10;
    cv::rectangle(frame, cv::Point(bx, by), cv::Point(bx + bw, by + bh),
                  cv::Scalar(0, 0, 0), cv::FILLED);
    cv::rectangle(frame, cv::Point(bx, by), cv::Point(bx + bw, by + bh),
                  cv::Scalar(100, 100, 100), 1);

    std::ostringstream ss;
    ss << std::fixed << std::setprecision(1);
    ss << "Alpha: " << pitch_;
    cv::putText(frame, ss.str(), cv::Point(bx + 8, by + 22),
                cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 255, 0), 1);
    ss.str(""); ss << "Beta: " << yaw_;
    cv::putText(frame, ss.str(), cv::Point(bx + 8, by + 42),
                cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 255, 0), 1);
    ss.str(""); ss << "Zoom: " << zoom_ << "x";
    cv::putText(frame, ss.str(), cv::Point(bx + 8, by + 62),
                cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 255, 0), 1);
}

} // namespace fusion
